### PR TITLE
Server improvements

### PR DESCRIPTION
**Pull request by scohen**
* handled encoding / decoding in the protocol
  * Cleaned up protocol negotiation
  * Server and client both use Codecs and Compressors

### CODE REVIEW

**Comment by scohen** [comment]
@jhowarth @jhgg @vishnevskiy  ping.

**Comment by scohen** [comment]
FYI, I benchmarked this and the old code via the go client. The old code did ~105-108k qps. This does ~130-140qps
Right now we're bottlenecked at spawning processes and handling DOWN messages.

File: ex/loqui/lib/loqui/cowboy_protocol.ex [hunk 1]
```diff
@@ -165,20 +209,21 @@ defmodule Loqui.CowboyProtocol do
 
   @spec handle_request(tuple, state) :: {:ok, state} | {:shutdown, atom}
   defp handle_request({:hello, _flags, version, encodings, compressions}, %{ping_interval: ping_interval, supported_encodings: supported_encodings, supported_compressions: supported_compressions}=state) do
-    encoding = choose_encoding(supported_encodings, encodings)
-    compression = choose_compression(supported_compressions, compressions)
-
+    codec = choose_encoding(supported_encodings, encodings)
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/cowboy_protocol.ex]: [nit] Might want to change everything to be called `codec` for consistency.

Comment (scohen) [reply ex/loqui/lib/loqui/cowboy_protocol.ex]: Agreed, but I think these are actually encodings, as they're just strings sent from the client. I'll update the struct's `:encoding` and `:compression` fields as well as the name of this function to `choose_codec` and `choose_compressor`.

**Pull request merged**